ports/stm32/mpbthciport.c

```diff
@@ -35,23 +36,59 @@
 
 #if MICROPY_PY_BLUETOOTH
 
-#define DEBUG_printf(...) // printf(__VA_ARGS__)
+#define DEBUG_printf(...) // printf("mpbthciport.c: " __VA_ARGS__)
 
 uint8_t mp_bluetooth_hci_cmd_buf[4 + 256];
 
 // Must be provided by the stack bindings (e.g. mpnimbleport.c or mpbtstackport.c).
-extern void mp_bluetooth_hci_poll(void);
+extern bool mp_bluetooth_hci_poll_uart(void);
+extern bool mp_bluetooth_hci_poll_events(void);
+
+#if MICROPY_PY_BLUETOOTH_USE_RINGBUFFER
 
 // Hook for pendsv poller to run this periodically every 128ms
 #define BLUETOOTH_HCI_TICK(tick) (((tick) & ~(SYSTICK_DISPATCH_NUM_SLOTS - 1) & 0x7f) == 0)
 
+STATIC void mp_bluetooth_hci_poll() {
+    mp_bluetooth_hci_poll_uart();
+    mp_bluetooth_hci_poll_events();
+}
+
 // Called periodically (systick) or directly (e.g. uart irq).
 void mp_bluetooth_hci_poll_wrapper(uint32_t ticks_ms) {
     if (ticks_ms == 0 || BLUETOOTH_HCI_TICK(ticks_ms)) {
         pendsv_schedule_dispatch(PENDSV_DISPATCH_BLUETOOTH_HCI, mp_bluetooth_hci_poll);
     }
 }
 
+void mp_bluetooth_on_hci_irq(void) {
+    // New HCI data, schedule mp_bluetooth_hci_poll via PENDSV to make the stack handle it.
+    mp_bluetooth_hci_poll_wrapper(0);
+}
+
+#else // ! MICROPY_PY_BLUETOOTH_USE_RINGBUFFER
+
+STATIC mp_obj_t hci_poll(mp_obj_t none_in);
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(hci_poll_obj, hci_poll);
+
+STATIC bool hci_scheduled = false;
+
+STATIC mp_obj_t hci_poll(mp_obj_t none_in) {
+    (void)none_in;
+    hci_scheduled = false;
+    if (mp_bluetooth_hci_poll_uart() && mp_bluetooth_hci_poll_events()) {
```

Comment:
Is it true that these two functions effectively always return `true` if ble is started/active? 
If so, once bluetooth is started `hci_poll()` will always continually re-schedule itself?

Reply:
Until `ble.active(0)` is called, yes.

There are likely some efficiency improvements to explore in the future along the lines of only schedule when something is added to the event/callout queue or when UART data arrives.

Reply:
yep sounds good, just wanted to be sure I was following it correctly. I agree, once this is known to be stable, we can look at selective scheduling (on uart / event / gat.write() calls / etc).